File: pages/swissborg.js
import { By } from 'selenium-webdriver';
import * as NumFormat from '../utils/numberFormatter.js';
import { handlerError } from '../utils/telegram.js';

// Click for accept cookie in Swissborg.
export const acceptCookieSwissborg = async (driver, maxLoop) => {
  try {
    let cookieButtons = undefined;
    while(!cookieButtons) {
      cookieButtons = await driver.findElements(By.className('cookieBanner__SButton-sc-190qymo-5'));
      if(cookieButtons) await cookieButtons[1].click();

      if(maxLoop === 0) throw new Error('Nb loop max for cookie button.'); 
      maxLoop--;
    }
  } catch(e) {
    console.error('Error with cookie button :' + e);
    throw new Error('Error with cookie button :' + e);
  }
}

// Get marketCap BORG.
export const getMarketCapBorg = async (borgMetrics, driver, maxLoop) => {
  try {
    while(!borgMetrics.marketCap || borgMetrics.marketCap === 'N/A' || borgMetrics.marketCap === '') {
      const marketCap = await driver.findElement(By.css('.cell-3 > p'));
      borgMetrics.marketCap = await marketCap.getText();
      if(maxLoop === 0) throw new Error('Nb loop max in getMarketCapBorg.');
      maxLoop--;
    }

    maxLoop = 5;
    console.log('MarketCap BORG is acquired.');
  } catch(e) {
    await handlerError(e, driver, 'Error to get marketCap BORG :');
  }
}

// Get prenium user BORG.
export const getPremiumUserBorg = async (borgMetrics, driver, maxLoop) => {
  try {
    while(!borgMetrics.premiumUser) {
      await driver.actions()
      .scroll(0, 0, 0, 300)
      .perform()
      await new Promise(resolve => setTimeout(resolve, 2000));
      const premiumUser = await driver.findElement(By.className('stat-0'));
      borgMetrics.premiumUser = await premiumUser.getText();
      if(maxLoop === 0) throw new Error('Nb loop max in getPremiumUserBorg.');
      maxLoop--;
    }

    maxLoop = 5;
    console.log('Nb premium user BORG is acquired.');
  } catch(e) {
    await handlerError(e, driver, 'Error to get nb premium user BORG :');
  }
}

// Get BORG blocked by user.
export const getBorgLockByPremium = async (borgMetrics, driver, maxLoop) => {
  try {
    while(!borgMetrics.borgLockByPremium) {
      const borgLockByPremium = await driver.findElement(By.className('stat-2'));
      borgMetrics.borgLockByPremium = await borgLockByPremium.getText();

      if(maxLoop === 0) throw new Error('Nb loop max in getBorgLockByPremium.', true);
      maxLoop--;
    }

    maxLoop = 5;
    console.log('Nb BORG lock is acquired.');
  } catch(e) {
    await handlerError(e, driver, 'Error to get nb BORG lock :');
  }
}

// Get supply in circulation.
export const getSupplyCirculationBorg = async (borgMetrics, driver, maxLoop) => {
  try {
    while(!borgMetrics.supplyCirculation) {
      const supplyCirculation = await driver.findElement(By.css('.cell-4 > p'));
      borgMetrics.supplyCirculation = await supplyCirculation.getText();
      borgMetrics.supplyCirculation = borgMetrics.supplyCirculation.split('S')[0];

      if(maxLoop === 0) throw new Error('Nb loop max in getSupplyCirculationBorg.');
      maxLoop--;
    }

    maxLoop = 5;
    console.log('Nb supply in circulation is acquired.');
  } catch(e) {
    await handlerError(e, driver, 'Error to get nb supply in circulation BORG :');
  }
}

// Get AUM BORG.
export const getAumBorg = async (borgMetrics, driver, maxLoop) => {
  try {
    while(!borgMetrics.aum) {
      await driver.actions()
      .scroll(0, 0, 0, 1000)
      .perform()
      await new Promise(resolve => setTimeout(resolve, 2000));

      const listData = await driver.findElements(By.className('bjTDmW'));
      borgMetrics.aum = await listData[2].getText();

      if(maxLoop === 0) throw new Error('Nb loop max in getAumBorg.');
      maxLoop--;
    }

    maxLoop = 5;
    console.log('AUM is acquired.');
  } catch(e) {
    await handlerError(e, driver, 'Error to get AUM BORG :');
  }
}

// Get verify user.
export const getUserVerify = async (borgMetrics, driver, maxLoop) => {
  try {
    while(!borgMetrics.userVerify) {
      const listData = await  await driver.findElements(By.className('bjTDmW'));
      borgMetrics.userVerify = await listData[1].getText();

      if(maxLoop === 0) throw new Error('Nb loop max in getUserVerify.');
      maxLoop--;
    }

    maxLoop = 5;
    console.log('Verify user is acquired.');
  } catch(e) {
    await handlerError(e, driver, 'Error to get verify user BORG :');
  }
}

// Get nb borg lock for governance.
export const getBorgLockForGovernance = async (borgMetrics, driver, maxLoop) => {
  try {
    while(!borgMetrics.borgLockForGovernance) {
      const title = await driver.findElement(By.className('iCCdso'));
      await driver.actions()
      .scroll(0, 0, 0, 0, title)
      .perform()
      await new Promise(resolve => setTimeout(resolve, 2000));

      const borgLockForGovernance = await driver.findElements(By.className('SZCGg'));
      borgMetrics.borgLockForGovernance = await borgLockForGovernance[7].getText(); // Return error or wrong value in DEV but not in PROD. Put 2 in DEV

      if(maxLoop === 0) throw new Error('Nb loop max in getBorgLockForGovernance.');
      maxLoop--;
    }

    maxLoop = 5;
    console.log('Borg lock for governance is acquired.');
  } catch(e) {
    await handlerError(e, driver, 'Error to get borg lock for governance :');
  }
}

// Get circulating Borg.
export const getCirculatingBorg = async (borgMetrics, driver, maxLoop) => {
  try {
    while(!borgMetrics.circulatingBorg) {
      const circulatingBorg = await driver.findElements(By.className('SZCGg'));
      borgMetrics.circulatingBorg = await circulatingBorg[5].getText(); // Return error or wrong value in DEV but not in PROD. Put 0 in DEV

      if(maxLoop === 0) throw new Error('Nb loop max in getCirculatingBorg.');
      maxLoop--;
    }

    maxLoop = 5;
    console.log('Circulating Borg is acquired.');
  } catch(e) {
    await handlerError(e, driver, 'Error to get circulating Borg :');
  }
}

// Calcul difference between old value and new value.
export const calculVariation = (borgMetrics, oldBorgMetrics, variationBorgMetrics) => {
  const marketCap = NumFormat.convertNumberForCalcul(borgMetrics.marketCap);
  const oldMarketCap = NumFormat.convertNumberForCalcul(oldBorgMetrics.marketCap);
  const supplyCirculation = NumFormat.convertNumberForCalcul(borgMetrics.supplyCirculation);
  const oldSupplyCirculation = NumFormat.convertNumberForCalcul(oldBorgMetrics.supplyCirculation);
  const volumeCoinGecko = NumFormat.convertNumberForCalcul(borgMetrics.volumeCoinGecko);
  const oldVolumeCoinGecko = NumFormat.convertNumberForCalcul(oldBorgMetrics.volumeCoinGecko);
  const userVerify = NumFormat.convertNumberForCalcul(borgMetrics.userVerify);
  const oldUserVerify = NumFormat.convertNumberForCalcul(oldBorgMetrics.userVerify);
  const premiumUser = NumFormat.convertNumberForCalcul(borgMetrics.premiumUser);
  const oldPremiumUser = NumFormat.convertNumberForCalcul(oldBorgMetrics.premiumUser);
  const borgLockByPremium = NumFormat.convertNumberForCalcul(borgMetrics.borgLockByPremium);
  const oldBorgLockByPremium = NumFormat.convertNumberForCalcul(oldBorgMetrics.borgLockByPremium);
  const aum = NumFormat.convertNumberForCalcul(borgMetrics.aum);
  const oldAum = NumFormat.convertNumberForCalcul(oldBorgMetrics.aum);
  const borgLockForGovernance = NumFormat.convertNumberForCalcul(borgMetrics.borgLockForGovernance);
  const oldBorgLockForGovernance = NumFormat.convertNumberForCalcul(oldBorgMetrics.borgLockForGovernance);
  const circulatingBorg = NumFormat.convertNumberForCalcul(borgMetrics.circulatingBorg);
  const oldCirculatingBorg = NumFormat.convertNumberForCalcul(oldBorgMetrics.circulatingBorg);

  // Percent
  variationBorgMetrics.value = borgMetrics.value && oldBorgMetrics.value ? ((Number(borgMetrics.value) - Number(oldBorgMetrics.value)) / Number(oldBorgMetrics.value) * 100).toFixed(2) : 'N/A';
  variationBorgMetrics.aum = ((aum - oldAum) / oldAum * 100).toFixed(2);
  variationBorgMetrics.marketCap = ((marketCap - oldMarketCap) / oldMarketCap * 100).toFixed(2);
  variationBorgMetrics.volumeCoinGecko = ((volumeCoinGecko - oldVolumeCoinGecko) / oldVolumeCoinGecko * 100).toFixed(2);

  variationBorgMetrics.userVerify = userVerify - oldUserVerify;
  variationBorgMetrics.premiumUser = premiumUser - oldPremiumUser;
  variationBorgMetrics.borgLockByPremium = borgLockByPremium - oldBorgLockByPremium;
  variationBorgMetrics.supplyCirculation = supplyCirculation - oldSupplyCirculation;
  variationBorgMetrics.rank = borgMetrics.rank && oldBorgMetrics.rank ? Number(borgMetrics.rank) - Number(oldBorgMetrics.rank) : 'N/A';
  variationBorgMetrics.borgLockForGovernance = borgLockForGovernance > 0 && oldBorgLockForGovernance > 0 ? borgLockForGovernance - oldBorgLockForGovernance : 'N/A';
  variationBorgMetrics.circulatingBorg = circulatingBorg > 0 && oldCirculatingBorg > 0 ? circulatingBorg - oldCirculatingBorg : 'N/A';
}

File: pages/swissborg.py
import time
from playwright.async_api import Page
import utils.numberFormatter as NumFormat
from utils.telegram import handler_error

# Click for accept cookie in Swissborg.
async def accept_cookie_swissborg(page: Page, max_loop: int):
  try:
    cookie_button = None
    while cookie_button is None:
      cookie_button = page.locator('.cookieBanner__SButton-sc-190qymo-5').last
      if cookie_button: await cookie_button.click()

      if max_loop == 0: raise Exception('Nb loop max for cookie button.')
      max_loop = max_loop - 1
  except Exception as e:
    print(f'Error with cookie button : {e}')
    # raise Exception(f'Error with cookie button : {e}') from e

# Get marketCap BORG.
async def get_market_cap_borg(borg_metrics: dict, page: Page, max_loop: int):
  try:
    while borg_metrics['marketCap'] == '' or borg_metrics['marketCap'] == 'N/A' or borg_metrics['marketCap'] == '':
      market_cap = page.locator('.cell-3').get_by_role('paragraph')
      borg_metrics['marketCap'] = await market_cap.text_content()
      if max_loop == 0: raise Exception('Nb loop max in get_market_cap_borg.')
      max_loop = max_loop - 1

    max_loop = 5
    print('MarketCap BORG is acquired.')
  except Exception as e:
    await handler_error(e, page, 'Error to get marketCap BORG :')

# Get prenium user BORG.
async def get_premium_user_borg(borg_metrics: dict, page: Page, max_loop: int):
  try:
    while borg_metrics['premiumUser'] == '':
      await page.locator('.AHLJo').scroll_into_view_if_needed()
      time.sleep(2)

      premium_user = page.locator('.stat-0')
      borg_metrics['premiumUser'] = await premium_user.text_content()
      if max_loop == 0: raise Exception('Nb loop max in get_premium_user_borg.')
      max_loop = max_loop - 1

    max_loop = 5
    print('Nb premium user BORG is acquired.')
  except Exception as e:
    await handler_error(e, page, 'Error to get nb premium user BORG :')

# Get BORG blocked by user.
async def get_borg_lock_by_premium(borg_metrics: dict, page: Page, max_loop: int):
  try:
    while borg_metrics['borgLockByPremium'] == '':
      time.sleep(1)
      borg_lock_by_premium = page.locator('.stat-2')
      borg_metrics['borgLockByPremium'] = await borg_lock_by_premium.text_content()

      if max_loop == 0: raise Exception('Nb loop max in get_borg_lock_by_premium.', True)
      max_loop = max_loop - 1

    max_loop = 5
    print('Nb BORG lock is acquired.')
  except Exception as e:
    await handler_error(e, page, 'Error to get nb BORG lock :')

# Get supply in circulation.
async def get_supply_circulation_borg(borg_metrics: dict, page: Page, max_loop: int):
  try:
    while borg_metrics['supplyCirculation'] == '':
      supply_circulation = page.locator('.cell-4').get_by_role('paragraph')
      borg_metrics['supplyCirculation'] = await supply_circulation.text_content()
      borg_metrics['supplyCirculation'] = borg_metrics['supplyCirculation'].split('S')[0]

      if max_loop == 0: raise Exception('Nb loop max in get_supply_circulation_borg.')
      max_loop = max_loop - 1

    max_loop = 5
    print('Nb supply in circulation is acquired.')
  except Exception as e:
    await handler_error(e, page, 'Error to get nb supply in circulation BORG :')

# Get AUM BORG.
async def get_aum_borg(borg_metrics: dict, page: Page, max_loop: int):
  try:
    while borg_metrics['aum'] == '':
      await page.locator('h2').last.scroll_into_view_if_needed()

      list_data = page.locator('.bjTDmW')
      borg_metrics['aum'] = await list_data.locator('nth=2').text_content()

      if max_loop == 0: raise Exception('Nb loop max in get_aum_borg.')
      max_loop = max_loop - 1

    max_loop = 5
    print('AUM is acquired.')
  except Exception as e:
    await handler_error(e, page, 'Error to get AUM BORG :')

# Get verify user.
async def get_user_verify(borg_metrics: dict, page: Page, max_loop: int):
  try:
    while borg_metrics['userVerify'] == '':
      list_data = page.locator('.bjTDmW')
      borg_metrics['userVerify'] = await list_data.locator('nth=1').text_content()

      if max_loop == 0: raise Exception('Nb loop max in get_user_verify.')
      max_loop = max_loop - 1

    max_loop = 5
    print('Verify user is acquired.')
  except Exception as e:
    await handler_error(e, page, 'Error to get verify user BORG :')

# Get nb borg lock for governance.
async def get_borg_lock_for_governance(borg_metrics: dict, page: Page, max_loop: int):
  try:
    while borg_metrics['borgLockForGovernance'] == '':
      await page.locator('.iMYwiA').scroll_into_view_if_needed()
      time.sleep(2)

      borg_lock_for_governance = page.locator('.SZCGg')
      borg_metrics['borgLockForGovernance'] = await borg_lock_for_governance.locator('nth=7').text_content()

      if max_loop == 0: raise Exception('Nb loop max in get_borg_lock_for_governance.')
      max_loop = max_loop - 1

    max_loop = 5
    print('Borg lock for governance is acquired.')
  except Exception as e:
    await handler_error(e, page, 'Error to get borg lock for governance :')

# Get circulating Borg.
async def get_circulating_borg(borg_metrics: dict, page: Page, max_loop: int):
  try:
    while borg_metrics['circulatingBorg'] == '':
      circulating_borg = page.locator('.SZCGg')
      borg_metrics['circulatingBorg'] = await circulating_borg.locator('nth=5').text_content()

      if max_loop == 0: raise Exception('Nb loop max in get_circulating_borg.')
      max_loop = max_loop - 1

    max_loop = 5
    print('Circulating Borg is acquired.')
  except Exception as e:
    await handler_error(e, page, 'Error to get circulating Borg :')

# Calcul difference between old value and new value.
def calcul_variation(borg_metrics: dict, old_borg_metrics: dict, variation_borg_metrics: dict):
  market_cap = NumFormat.convert_number_for_calcul(borg_metrics['marketCap'])
  old_market_cap = NumFormat.convert_number_for_calcul(old_borg_metrics['marketCap'])
  supply_circulation = NumFormat.convert_number_for_calcul(borg_metrics['supplyCirculation'])
  old_supply_circulation = NumFormat.convert_number_for_calcul(old_borg_metrics['supplyCirculation'])
  volume_coingecko = NumFormat.convert_number_for_calcul(borg_metrics['volumeCoinGecko'])
  old_volume_coingecko = NumFormat.convert_number_for_calcul(old_borg_metrics['volumeCoinGecko'])
  user_verify = NumFormat.convert_number_for_calcul(borg_metrics['userVerify'])
  old_user_verify = NumFormat.convert_number_for_calcul(old_borg_metrics['userVerify'])
  premium_user = NumFormat.convert_number_for_calcul(borg_metrics['premiumUser'])
  old_premium_user = NumFormat.convert_number_for_calcul(old_borg_metrics['premiumUser'])
  borg_lock_by_premium = NumFormat.convert_number_for_calcul(borg_metrics['borgLockByPremium'])
  old_borg_lock_by_premium = NumFormat.convert_number_for_calcul(old_borg_metrics['borgLockByPremium'])
  aum = NumFormat.convert_number_for_calcul(borg_metrics['aum'])
  old_aum = NumFormat.convert_number_for_calcul(old_borg_metrics['aum'])
  borg_lock_for_governance = NumFormat.convert_number_for_calcul(borg_metrics['borgLockForGovernance'])
  old_borg_lock_for_governance = NumFormat.convert_number_for_calcul(old_borg_metrics['borgLockForGovernance'])
  circulating_borg = NumFormat.convert_number_for_calcul(borg_metrics['circulatingBorg'])
  old_circulating_borg = NumFormat.convert_number_for_calcul(old_borg_metrics['circulatingBorg'])

  # Percent
  variation_borg_metrics['value'] = round(((float(borg_metrics['value']) - float(old_borg_metrics['value'])) / float(old_borg_metrics['value']) * 100), 2) if borg_metrics['value'] and old_borg_metrics['value'] else 'N/A'
  variation_borg_metrics['aum'] = round(((aum - old_aum) / old_aum * 100), 2)
  variation_borg_metrics['marketCap'] = round(((market_cap - old_market_cap) / old_market_cap * 100), 2)
  variation_borg_metrics['volumeCoinGecko'] = round(((volume_coingecko - old_volume_coingecko) / old_volume_coingecko * 100), 2)

  variation_borg_metrics['userVerify'] = user_verify - old_user_verify
  variation_borg_metrics['premiumUser'] = premium_user - old_premium_user
  variation_borg_metrics['borgLockByPremium'] = borg_lock_by_premium - old_borg_lock_by_premium
  variation_borg_metrics['supplyCirculation'] = supply_circulation - old_supply_circulation
  variation_borg_metrics['rank'] = int(borg_metrics['rank']) - int(old_borg_metrics['rank']) if borg_metrics['rank'] and old_borg_metrics['rank'] else 'N/A'
  variation_borg_metrics['borgLockForGovernance'] = borg_lock_for_governance - old_borg_lock_for_governance if borg_lock_for_governance > 0 and old_borg_lock_for_governance > 0 else 'N/A'
  variation_borg_metrics['circulatingBorg'] = circulating_borg - old_circulating_borg if circulating_borg > 0 and old_circulating_borg > 0 else 'N/A'